src/modules/kvs-watch/kvs-watch.c

```diff
@@ -311,6 +321,73 @@ static int handle_compare_response (flux_t *h,
     return 0;
 }
 
+static int handle_append_response (flux_t *h,
+                                    struct watcher *w,
+                                    json_t *val)
+{
+    if (!w->responded) {
+        /* this is the first response case, store the first response
+         * info.  This is here b/c initial response could have been
+         * ENOENT case */
+        if (treeobj_decode_val (val,
+                                NULL,
+                                &w->append_offset) < 0) {
+            flux_log_error (h, "%s: treeobj_decode_val", __FUNCTION__);
+            return -1;
+        }
+
+        if (flux_respond_pack (h, w->request, "{ s:O }", "val", val) < 0) {
+            flux_log_error (h, "%s: flux_respond_pack", __FUNCTION__);
+            return -1;
+        }
+    }
+    else {
+        json_t *new_val = NULL;
+        void *new_data = NULL;
+        int new_offset;
+
+        if (treeobj_decode_val (val,
+                                &new_data,
+                                &new_offset) < 0) {
+            flux_log_error (h, "%s: treeobj_decode_val", __FUNCTION__);
+            return -1;
+        }
+
+        /* check length to determine if append actually happened, note
+         * that zero length append is legal
+         *
+         * Note that this check does not ensure that the key was not
+         * "fake" appended to.  i.e. the key overwritten with data
+         * longer than the original.
+         */
+        if (new_offset < w->append_offset) {
+            free (new_data);
+            errno = EINVAL;
+            return -1;
+        }
+
+        if (!(new_val = treeobj_create_val (new_data + w->append_offset,
+                                            new_offset - w->append_offset))) {
+            free (new_data);
+            return -1;
+        }
+
+        free (new_data);
+        w->append_offset = new_offset;
+
+        if (flux_respond_pack (h, w->request, "{ s:O }", "val", new_val) < 0) {
+            json_decref (new_val);
+            flux_log_error (h, "%s: flux_respond_pack", __FUNCTION__);
+            return -1;
+        }
+
+        json_decref (new_val);
+    }
+
+    w->responded = true;
```

Comment:
set inside !w->responded conditional leg?

Reply:
ahh, i copied this from the ```handle_compare_response()``` function, should move it in there too.